### ISSUE TITLE
Regression in 1.1.218 involving recursive types in tuples

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#2998]
The following false-negative occurs in pyright 1.1.218

```python
from typing import Union, Tuple

A = Union[str, int]
B = Union[A, Tuple["B", ...]]

class M:
    pass

def g(x: B):
    return


# (pyright 1.1.217) error: Argument of type "tuple[Literal[1], M]" cannot be assigned to parameter "x" of type "B"
# (pyright 1.1.218) no error
g((1, M())) 
```

**Comment by erictraut** [comment]
Thanks for catching this. A bug fix that I introduced in 1.1.218 uncovered a latent bug that affected the evaluation of tuples, and this particular case triggers it. I also didn't have any unit tests in place for recursive tuples, so the tests didn't detect this.

I've fixed the bug and added a new unit test based on your example. The fix will be in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.219, which I just published. It will also be in the next release of pylance.
